crates/ruff/src/rules/ruff/rules/quadratic_list_summation.rs

```diff
@@ -0,0 +1,114 @@
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, Arguments, Expr};
+use ruff_text_size::TextRange;
+
+use crate::{checkers::ast::Checker, registry::Rule};
+
+/// ## What it does
+/// Avoid quadratic
+///
+/// ## Why is this bad?
+/// Quadratic list summation is slower than other methods of list concatenation.
+/// A list comprehension can perform the same operation but in a much shorter time period.
+///
+/// ## Example
+/// ```python
+/// lists = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
+/// joined = sum(lists, [])
+/// ```
+///
+/// Use instead:
+/// ```python
+/// lists = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
+/// joined = [el for list in lists for el in list]
+/// ```
+#[violation]
+pub struct QuadraticListSummation;
+
+impl AlwaysAutofixableViolation for QuadraticListSummation {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Replace quadratic list summation with list comprehension")
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Convert to list comprehension")
+    }
+}
+
+/// RUF017
+pub(crate) fn quadratic_list_summation(checker: &mut Checker, call: &ast::ExprCall) {
+    let ast::ExprCall {
+        func,
+        arguments,
+        range,
+    } = call;
+
+    if !func_is_builtin(func, "sum", checker) {
+        return;
+    }
+
+    if verify_start_arg(arguments, checker).is_none() {
+        return;
+    }
+
+    let Arguments { args, .. } = arguments;
+    let Some(list_string_repr) = args
+        .first()
```

Comment:
nit: use the following pattern for slightly more straightforward code
```suggestion
    let Some(first) = args.first() else {
        return;
    }

    match first {
        Expr::Name(ast::ExprName { id, .. }) => ...,
        Expr::List(ast::ExprList { range, .. }) => ...,
        _ => return,
    }
```